File: domain-driven-design/go/domain/order.go
package domain

import "fmt"

// 订单状态枚举
type OrderStatus int

const (
  CREATED OrderStatus = iota
  CANCELED
)

// 订单结构体
type Order struct {
  ID           int
  CustomerName string
  Amount       float64
  Status       OrderStatus
}

// 创建订单
func NewOrder(id int, customerName string, amount float64) *Order {
  if amount < 0 {
    fmt.Println("订单金额不能为负数")
    return nil
  }
  return &Order{
    ID:           id,
    CustomerName: customerName,
    Amount:       amount,
    Status:       CREATED,
  }
}

// 取消订单
func (o *Order) Cancel() {
  if o.Status == CREATED {
    o.Status = CANCELED
    fmt.Printf("订单 ID %d 已取消\n", o.ID)
  } else {
    fmt.Printf("订单 ID %d 已经被取消，无法重复操作\n", o.ID)
  }
}

// 显示订单信息
func (o *Order) Display() {
  fmt.Printf("订单 ID: %d\n客户名称: %s\n订单金额: %.2f\n订单状态: %s\n",
    o.ID, o.CustomerName, o.Amount, o.getStatusString())
}

func (o *Order) getStatusString() string {
  switch o.Status {
  case CREATED:
    return "已创建"
  case CANCELED:
    return "已取消"
  default:
    return "未知状态"
  }
}

File: domain-driven-design/c/domain/order.c
// 领域层（Domain Layer） - order.c
#include "order.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// 创建订单
Order *createOrder(int id, const char *customerName, double amount)
{
    if (amount < 0)
    {
        printf("订单金额不能为负数\n");
        return NULL;
    }

    Order *order = (Order *)malloc(sizeof(Order));
    if (!order)
    {
        printf("内存分配失败\n");
        return NULL;
    }

    order->id = id;
    strncpy(order->customerName, customerName, MAX_NAME_LENGTH - 1);
    order->customerName[MAX_NAME_LENGTH - 1] = '\0'; // 确保字符串以 '\0' 结尾
    order->amount = amount;
    order->status = CREATED; // 订单默认状态

    return order;
}

// 取消订单
void cancelOrder(Order *order)
{
    if (!order)
    {
        printf("订单不存在\n");
        return;
    }

    if (order->status == CREATED)
    {
        order->status = CANCELED;
        printf("订单 ID %d 已取消\n", order->id);
    }
    else
    {
        printf("订单 ID %d 已经取消，无法重复操作\n", order->id);
    }
}

// 释放订单
void freeOrder(Order *order)
{
    if (order)
    {
        free(order);
    }
}

// 显示订单信息
void displayOrderInfo(const Order *order)
{
    if (!order)
    {
        printf("无效订单\n");
        return;
    }

    printf("订单 ID: %d\n", order->id);
    printf("客户名称: %s\n", order->customerName);
    printf("订单金额: %.2f\n", order->amount);
    printf("订单状态: %s\n", order->status == CREATED ? "已创建" : "已取消");
}
